core/trino-main/src/test/java/io/trino/sql/query/TestListagg.java

```diff
@@ -0,0 +1,301 @@
+    public void testListaggQueryWithDistinct()
+    {
+        assertThat(assertions.query(
+                "SELECT listagg( DISTINCT value, ',') WITHIN GROUP (ORDER BY value) " +
+                        "FROM (VALUES  'a', 'b', 'a', 'b', 'c', 'd', 'd', 'a', 'd', 'b', 'd') t(value)"))
+                .matches("VALUES (VARCHAR 'a,b,c,d')");
+    }
+
+    @Test
+    public void testListaggQueryWithMultipleValuesWithDefaultSeparator()
+    {
+        assertThat(assertions.query(
+                "SELECT listagg(value) WITHIN GROUP (ORDER BY value) " +
+                        "FROM (VALUES 'a', 'bb', 'ccc', 'dddd') t(value) "))
+                .matches("VALUES (VARCHAR 'abbcccdddd')");
+    }
+
+    @Test
+    public void testListaggQueryWithOrdering()
+    {
+        assertThat(assertions.query(
+                "SELECT LISTAGG(value, ',') WITHIN GROUP (ORDER BY value) " +
+                        "FROM (VALUES 'a', 'ccc', 'dddd', 'bb') t(value) "))
+                .matches("VALUES (VARCHAR 'a,bb,ccc,dddd')");
+    }
+
+    @Test
+    public void testListaggQueryWithOrderingAndGrouping()
+    {
+        assertThat(assertions.query("SELECT id, LISTAGG(value, ',') WITHIN GROUP (ORDER BY value) " +
+                "          FROM (VALUES " +
+                "                   (1, 'a'), " +
+                "                   (1, 'b'), " +
+                "                   (2, 'd'), " +
+                "                   (2, 'c') " +
+                "               ) t(id, value)" +
+                "          GROUP BY id" +
+                "          ORDER BY id"))
+                .matches("VALUES     " +
+                        "     (1, VARCHAR 'a,b')," +
+                        "     (2, VARCHAR 'c,d')");
+    }
+
+    @Test
+    public void testListaggQueryOverflowError()
+    {
+        String tooLargeValue = StringUtils.repeat("a", DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+        assertThatThrownBy(() -> assertions.query(
+                "SELECT LISTAGG(value, ',' ON OVERFLOW ERROR) WITHIN GROUP (ORDER BY value) " +
+                        "FROM (VALUES '" + tooLargeValue + "','Trino') t(value) "))
+                .isInstanceOf(TrinoException.class)
+                .hasMessage("Concatenated string has the length 1048582 which is larger than the maximum output length 1048576")
+                .matches(throwable -> ((TrinoException) throwable).getErrorCode() == EXCEEDED_FUNCTION_MEMORY_LIMIT.toErrorCode());
+    }
+
+    @Test
+    public void testListaggQueryOverflowTruncateWithoutCountAndWithoutOverflowFiller()
+    {
+        String largeValue = StringUtils.repeat("a", DEFAULT_MAX_PAGE_SIZE_IN_BYTES - 6);
+        assertThat(assertions.query(
+                "SELECT LISTAGG(value, ',' ON OVERFLOW TRUNCATE WITHOUT COUNT) WITHIN GROUP (ORDER BY value) " +
+                        "FROM (VALUES '" + largeValue + "', 'trino', 'rocks') t(value) "))
+                .matches("VALUES (VARCHAR '" + largeValue + ",rocks...')");
```

Comment:
The output should end in `,rocks,...` (there's a missing separator between the last element and the filler)

Reply:
`largeValue` has `DEFAULT_MAX_PAGE_SIZE_IN_BYTES - 6` bytes 
`,rocks` has `6` bytes

My understanding was to a previous suggestion from your side was to either add the value or the separator only in full if they fit within the constraints of `MAX_OUTPUT_LENGTH`.
In the current test case, there's no more place available for the separator because the `MAX_OUTPUT_LENGTH` has been reached. As a consequence the filler is added at the end.

Reply:
Sorry if I wasn't clear before. My code above was incomplete with respect to this case. It's missing the addition of a separator before the filler is added.

In a nutshell,
* we should treat MAX_OUTPUT_LENGTH as a soft limit. It's ok if the size is exceeded by one entry's worth of space
* the output should be consistent with what's described in the SQL spec, which requires that the filler be added after a separator

Reply:
I've changed the logic to address the two points mentioned previously.